### PR TITLE
bug fix #145 #143 #135 #153

### LINKED ISSUE #135: normalize mishandles multi-word numbers
```python3
>>> parse.normalize("four thirty seven")
'4 30 7'
```

### LINKED ISSUE #143: bug in extract datetime - numbers not normalized
"2 minutes" works, "two minutes" does not

this is because numbers are not normalized

this both is and isnt a simple fix, text should just be normalized BUT

- normalize only replaces numbers up to 20
- i added a partial fix to this
- fix was reverted, see discussion here https://github.com/MycroftAI/mycroft-core/issues/1962

this should probably be split into two PRs/discussions, one adding the normalization in extract_datetime, other for fixing normalization

### LINKED ISSUE #145: bug - extract duration lowercases everything
extract_duration should not lower case the text in the remainder text, this might mess up comparisons with original string and simple replacements

### LINKED ISSUE #153: better ordinals handling
when working on #135 i realized we need better control over ordinals flag

- first is non ambiguous 
- second can be the time unit, not a number
- third and above can be a fraction
- things like "half" should not be normalized either, but parsed by extract_number

the same distinction might be useful for extract_number, therefore i propose

ordinals=True -> consider edge cases an integer
ordinals=False -> consider edge cases a fraction
ordinals=None -> ignore ambiguous words

things like 1st, 2nd, 3rd, 4th... are explicit numbers and should always be extracted

things like "whole", "half" or other spoken language quirks should also be optional

### PR DESCRIPTION
**Pull request by JarbasAl**
- case preservation in extract_duration
- extract_datetime number words support
- number replacement in normalize
- adds better support for ordinals

fixes #145
fixes #143 
fixes #135
fixed #153

### CODE REVIEW

File: lingua_franca/lang/parse_en.py [hunk 1]
```diff
@@ -622,8 +621,12 @@ def extract_datetime_en(text, anchorDate=None, default_time=None):
...
+        # normalize and lowercase utt  (replaces words with numbers=
+        s = normalize_en(s, remove_articles=False).lower()
+        # restore some decimal markers
+        s = s.replace("0.5", "half").replace("0.25", "quarter")
         # clean unneeded punctuation and capitalization among other things.
-        s = s.lower().replace('?', '').replace('.', '').replace(',', '') \
+        s = s.replace('?', '').replace('.', '').replace(',', '') \
             .replace(' the ', ' ').replace(' a ', ' ').replace(' an ', ' ') \
             .replace("o' clock", "o'clock").replace("o clock", "o'clock") \
             .replace("o ' clock", "o'clock").replace("o 'clock", "o'clock") \
```
Comment (ChanceNCounter) [inline lingua_franca/lang/parse_en.py:632]: OOS here, but surely this could be a regex looking for (invalid regex) `o*'*clock` (noting here because I just noticed)

File: lingua_franca/lang/parse_en.py [hunk 2]
```diff
@@ -45,7 +45,6 @@ def _convert_words_to_numbers_en(text, short_scale=True, ordinals=False):
         The original text, with numbers subbed in where appropriate.
 
     """
-    text = text.lower()
     tokens = tokenize(text)
```
Comment (ChanceNCounter) [inline lingua_franca/lang/parse_en.py:48]: Pros and cons vs. `tokens = [token.lower() for token in tokenize(text)]`?

**Review by ChanceNCounter — COMMENTED** [review]
Just the one question, and an out-of-scope observation that jumped out at me.

File: lingua_franca/lang/parse_en.py (hunk 2 above)
Comment (JarbasAl) [reply lingua_franca/lang/parse_en.py:48]: tokens are objects, not strings, and they should keep original string for phrase reconstruction

the .lower() operation is only relevant in comparisons and should NOT be propagated to the remainder text

File: lingua_franca/lang/parse_en.py (hunk 1 above)
Comment (JarbasAl) [reply lingua_franca/lang/parse_en.py:632]: out of scope, will be handled once my datetime rewrite is finished, essentially any changes to extract_datetime are currently a waste of time

File: test/test_parse.py [hunk 3]
```diff
@@ -221,27 +221,52 @@ def test_extract_duration_en(self):
...
+            extractedDate = extractedDate.strftime("%Y-%m-%d %H:%M:%S")
+            return [extractedDate, leftover]
+
+        def testExtract(text, expected_date, expected_leftover):
+            res = extractWithFormat(normalize(text))
+            self.assertEqual(res[0], expected_date, "for=" + text)
+            self.assertEqual(res[1], expected_leftover, "for=" + text)
+
+        testExtract("Set the ambush for half an hour",
+                    "2017-06-27 13:34:00", "set ambush")
```
Comment (JarbasAl) [inline test/test_parse.py:260]: NOTE: remainder is also lowercased here, opening a new issue for this elsewhere

File: test/test_parse.py [hunk 4]
```diff
@@ -697,15 +739,16 @@ def test_numbers(self):
         self.assertEqual(normalize("that's eighteen nineteen twenty"),
                          "that is 18 19 20")
         self.assertEqual(normalize("that's one nineteen twenty two"),
-                         "that is 1 19 20 2")
+                         "that is 1 19 22")
```
Comment (JarbasAl) [inline test/test_parse.py:742]: for these unittests we might want to throw a deprecation warning in case some downstream depends on this, relevant discussion https://github.com/MycroftAI/mycroft-core/issues/1962

File: test/test_parse.py (hunk 3 above)
Comment (JarbasAl) [reply test/test_parse.py:260]: documented here https://github.com/MycroftAI/lingua-franca/issues/147

**Comment by ChanceNCounter** [comment]
Wellp, I was just being obtuse with my only review comment =P

Tests passing, satisfies the issues, that DeprecationWarning is up to you. Looks good to me.

**Comment by JarbasAl** [comment]
id rather not throw the deprecation warning, i consider this a bug, you dont tell your users "we are delaying this fix until next release, work around it somehow"

**Comment by ChanceNCounter** [comment]
We could *raise* a DeprecationError, so at least stuff in the wild breaks... descriptively?

**Comment by ChanceNCounter** [comment]
Suppose that'd require you to put the correct behavior on a flag in the interim, which still isn't great...

File: test/test_parse.py [hunk 5]
```diff
@@ -71,20 +71,107 @@ def test_articles(self):
...
+                                        ordinals=None), False)
+
+        self.assertEqual(extract_number("one third of a cup"), 1.0 / 3.0)
+        self.assertEqual(extract_number("one third of a cup",
+                                        ordinals=True), 3)
+        self.assertEqual(extract_number("one third of a cup",
+                                        ordinals=None), 1)
+
+        # test plurals
+        # NOTE plurals are never considered ordinals, but also not
```
Comment (JarbasAl) [inline test/test_parse.py:120]: @ChanceNCounter @krisgesling does this note make sense?

File: test/test_parse.py [hunk 6]
```diff
@@ -71,20 +71,107 @@ def test_articles(self):
...
+                                        ordinals=True), 5)
+        self.assertEqual(extract_number("2 fifths",
+                                        ordinals=False), 2/5)
+        self.assertEqual(extract_number("2 fifths",
+                                        ordinals=None), 2)
+
+        self.assertEqual(extract_number("Twenty two and Three Fifths"), 22.6)
+
+        # TODO fix me - see issue #152
+        # these currently return 22 instead of 3
```
Comment (JarbasAl) [inline test/test_parse.py:134]: once #152 if fixed this test needs updated, tagging the issue here for reference

File: test/test_parse.py [hunk 7]
```diff
@@ -697,13 +831,15 @@ def test_numbers(self):
...
         self.assertEqual(normalize("that's one hundred"),
-                         "that is 1 hundred")
+                         "that is 100")
         self.assertEqual(normalize("that's one two twenty two"),
-                         "that is 1 2 20 2")
+                         "that is 1 2 22")
         self.assertEqual(normalize("that's one and a half"),
                          "that is 1 and half")
+
+        # TODO fix me
```
Comment (JarbasAl) [inline test/test_parse.py:842]: this test is failing and i could not figure out why yet, this is the main blocker for merging this PR currently, marking as draft until solved

File: test/test_parse.py (hunk 4 above)
Comment (JarbasAl) [reply test/test_parse.py:834]: i dont think we need a warning, if anything dowtreamns is depending on this behavior it is already utterly broken for n > 20 in general... we would have received bug reports already.

File: lingua_franca/lang/parse_en.py [hunk 8]
```diff
@@ -400,7 +409,7 @@ def _extract_whole_number_with_text_en(tokens, short_scale, ordinals):
                 current_val = val
 
         else:
-            if all([
+            if current_val and all([
```
Comment (JarbasAl) [inline lingua_franca/lang/parse_en.py:412]: this is an unrelated minor bug fix, during testing i ran into a case where current_val was None and int comparison inside all() would throw an exception

**Comment by JarbasAl** [comment]
eventually ordinals flag in extract number should become an Enum, for backwards compatibility this should not be changed yet but only in next major version

File: test/test_parse.py [hunk 9]
```diff
@@ -71,20 +71,105 @@ def test_articles(self):
...
+        self.assertEqual(extract_number("Twenty two and Three Fifths"), 22.6)
+
+        # TODO fix me - see issue #152
+        # these currently return 22 instead of 3
+        #self.assertEqual(extract_number("Twenty two and Three Fifths",
+        #                                ordinals=True), 3)
+
+        # test multiple ambiguous
+        self.assertEqual(extract_number("sixth third"),
+                         1 / 6 / 3)
```
Comment (krisgesling) [inline test/test_parse.py:140]: I don't think this one makes sense. 
"Sixth third" would probably refer to "the sixth instance of a third" which I guess means there's `2` all up.

File: test/test_parse.py (hunk 5 above)
Comment (ChanceNCounter) [reply test/test_parse.py:120]: Since you have to extract ordinals explicitly, and there's no such ordinal as The Three Fifth, it's probably sensible to avoid returning fractions in all cases.

On the other hand, it might make sense to call them ordinals, on the assumption that `extract_number("the thirds", ordinals=True)` will always be an input failure somewhere upstairs, like STT mishearing "the third."

File: test/test_parse.py (hunk 9 above)
Comment (ChanceNCounter) [reply test/test_parse.py:140]: Agreed. I dunno what should be returned here, but I don't think it should be cumulative.

Comment (JarbasAl) [reply test/test_parse.py:140]: this one has been around for ages, i just changed it to a different place so all related tests are grouped together

i dont think it makes a lot of sense either, but i'm declaring it out of scope since it isnt touched by this PR

File: test/test_parse.py [hunk 10]
```diff
@@ -71,20 +71,105 @@ def test_articles(self):
...
+        self.assertEqual(extract_number("this is the third test"),
+                         1.0 / 3.0)
+        self.assertEqual(extract_number("this is the third test",
+                                        ordinals=None), False)
+
+        self.assertEqual(extract_number("one third of a cup"), 1.0 / 3.0)
+        self.assertEqual(extract_number("one third of a cup",
+                                        ordinals=True), 3)
+        self.assertEqual(extract_number("one third of a cup",
+                                        ordinals=None), 1)
```
[the remark this answers is not in the document]
Comment (JarbasAl) [reply test/test_parse.py:117]: at some point i would like to replace this with an Enum, i did it this way for backwards compat only. but imho None should be the default in a next major release

File: test/test_parse.py [hunk 11]
```diff
@@ -71,20 +71,105 @@ def test_articles(self):
...
+        self.assertEqual(extract_number("remind me in a second",
+                                        ordinals=False), False)
+        self.assertEqual(extract_number("remind me in a second",
+                                        ordinals=None), False)
+
+        # test ambiguous ordinal/fractional
         self.assertEqual(extract_number("this is the third test",
                                         ordinals=True), 3.0)
+        self.assertEqual(extract_number("this is the third test"),
+                         1.0 / 3.0)
```
[the remark this answers is not in the document]
Comment (JarbasAl) [reply test/test_parse.py:109]: makes sense, will change :+1:

File: test/test_parse.py (hunk 5 above)
Comment (JarbasAl) [reply test/test_parse.py:120]: i was unsure of this one, for consistency i thought it made sense that ordinals=True would not parse fractions at all, the same way ordinals=False would only parse fractions.  But since we support explict ordinals (1st, 2nd, 3rd, 4th...) i'm not sure....

i guess its a matter of how explicit is this? what are the possible sources of ambiguity?

"the thirds" could very well mean 3 instead of 1/3 , so i don't think this should be considered an explicit fraction

Comment (JarbasAl) [reply test/test_parse.py:120]: i wonder if when ordinals=True we should consider the plural form an integer?

File: test/test_parse.py [hunk 12]
```diff
@@ -71,20 +71,105 @@ def test_articles(self):
...
+
+        # TODO fix me - see issue #152
+        # these currently return 22 instead of 3
+        #self.assertEqual(extract_number("Twenty two and Three Fifths",
+        #                                ordinals=True), 3)
+
+        # test multiple ambiguous
+        self.assertEqual(extract_number("sixth third"),
+                         1 / 6 / 3)
+        self.assertEqual(extract_number("sixth third", ordinals=True),
```
Comment (JarbasAl) [inline test/test_parse.py:141]: this one might hint at a deeper issue, since #152 should be breaking this test but isn't

File: test/test_parse.py [hunk 13]
```diff
@@ -671,6 +756,54 @@ def testExtract(text, expected_date, expected_leftover):
...
+        self.assertEqual(normalize("remind me to do something at twenty to two"),
+                         "remind me to do something at 20 to 2")
+
+        # TODO imperfect test, maybe should return 'my favorite numbers are 20 2',
+        #  let is pass for now since this is likely a STT issue if ever
+        #  encountered in the wild and is somewhat ambiguous, if this was
+        #  spoken by a human the result is what we expect, if in written form
+        #  it is ambiguous but could mean separate numbers
+        self.assertEqual(normalize('my favorite numbers are twenty 2'),
+                         'my favorite numbers are 22')
```
Comment (krisgesling) [inline test/test_parse.py:775]: may also be worth adding in where both numbers are digits 
> "my favorite numbers are 20 2"

Which I would expect would remain unchanged.

Comment (JarbasAl) [reply test/test_parse.py:781]: added a test for this, but it is also changed. This shares logic with extract_number and will be hard to fix without a bigger refactor, if you are ok with that i would like to leave it for a follow up PR

File: test/test_parse.py (hunk 9 above)
Comment (JarbasAl) [reply test/test_parse.py]: added a code comment for this one

**Comment by JarbasAl** [comment]
whats blocking this?

**Comment by ChanceNCounter** [comment]
Next in line after Catalan, let this block release IMO. Shouldn't be too bad. The will-be merge conflicts look like they should be click-fixes.

**Comment by ChanceNCounter** [comment]
squashed and pushed. speak soonish or forever hold your peace 🚀
